source/common/tls/context_config_impl.cc

```diff
@@ -447,6 +448,28 @@ ServerContextConfigImpl::ServerContextConfigImpl(
     session_timeout_ =
         std::chrono::seconds(DurationUtil::durationToSeconds(config.session_timeout()));
   }
+
+  TlsCertificateSelectorFactoryContextImpl provider_factory_context(api_, options_,
+                                                                    singleton_manager_);
+  if (config.common_tls_context().has_custom_tls_certificate_selector()) {
+    // If a custom tls context provider is configured, derive the factory from the config.
+    const auto& provider_config = config.common_tls_context().custom_tls_certificate_selector();
+    Ssl::TlsCertificateSelectorFactory* provider_factory =
+        &Config::Utility::getAndCheckFactory<Ssl::TlsCertificateSelectorFactory>(provider_config);
+    tls_certificate_selector_factory_cb_ = provider_factory->createTlsCertificateSelectorCb(
+        provider_config.typed_config(), provider_factory_context,
+        factory_context.messageValidationVisitor());
+  } else {
+    auto factory = Envoy::Config::Utility::getFactoryByName<Ssl::TlsCertificateSelectorFactory>(
```

Comment:
Copy the pattern for the handshaker; statically get the default factory so that there's no possible failure case here.

Reply:
It follow this commit for mobile: [server ssl context out of mobile build](https://github.com/envoyproxy/envoy/commit/7dca2c80d959a19fa165f16a82489ad2dc334ebf).

Reply:
But that was solving a different problem. This code is already inside the server context, so it's already compiled out for mobile. So please follow the pattern for the handshaker here, and do something like `handshaker_factory = HandshakerFactoryImpl::getDefaultHandshakerFactory();`.

Reply:
I had a try in [584a5bdd](https://github.com/envoyproxy/envoy/pull/32465/commits/584a5bddae9f53718c8d09c8d7f025938cf7890f#diff-89b1f3f3d469e5771ec6df3eb2375e210ac46a89c292386f18ffc76331caed9dR148), but with no luck.

After adding `server_context_lib` as `context_config_lib`'s deps, it leads two mobile test failure:

1. cannot use 'throw' with exceptions disabled
```
external/envoy/source/common/tls/server_context_impl.cc:107:5: error: cannot use 'throw' with exceptions disabled
    throw EnvoyException("Server TlsCertificates must have a certificate specified");
```
https://github.com/envoyproxy/envoy/actions/runs/9622716616/job/26546308242#step:8:429
This could be resolved by moving those exceptions into `creation_status`.

2. missing files in `test/performance:test_binary_size`
```
INFO: Analyzed target //test/performance:test_binary_size (557 packages loaded, 31300 targets configured).
INFO: Found 1 target...
ERROR: /root/.cache/bazel/_bazel_root/bb3efe6b13248f746491cbfc24ae2efe/external/envoy/source/common/tls/BUILD:215:17: Middleman _middlemen/@envoy_S_Ssource_Scommon_Stls_Cserver_Ucontext_Ulib-cc_library-compile failed: missing input file '@envoy//source/common/tls:default_tls_certificate_selector.h'
ERROR: /root/.cache/bazel/_bazel_root/bb3efe6b13248f746491cbfc24ae2efe/external/envoy/source/common/tls/BUILD:215:17: Middleman _middlemen/@envoy_S_Ssource_Scommon_Stls_Cserver_Ucontext_Ulib-cc_library-compile failed: missing input file '@envoy//source/common/tls:server_context_impl.h'
```
https://github.com/envoyproxy/envoy/actions/runs/9622716611/job/26546308513#step:8:434
seems this is unexpected for mobile?

maybe we can keep using factory registry? or any other suggestions? thanks.

Reply:
I think you can work around those issues by separating out the ServerContextConfig into it's own library, and not including that in mobile. cc @alyssawilk who's been working on removing stuff from mobile that isn't needed.

Reply:
yeah that'd be great to split out - it's on my list I just haven't gotten to it.  Suggest doing it as a standalone and merging ahead of this change.

Reply:
I'll work on that this afternoon

Reply:
#35050. Ideally I'd like to put the ClientContextConfigImpl into it's own file as well, but I'm short on time.

Reply:
Great, I'll have another try when #35050 is merged.